core/network/fabric-interop-cc/contracts/interop/manage_assets.go

```diff
@@ -0,0 +1,304 @@
+	} else {
+		log.Info(fmt.Printf("checkIfCorrectPreimage: preimage %s is not passed correctly.\n", preimage))
+		return false, nil
+	}
+	return true, nil
+}
+
+// ClaimAsset cc is used to record claim of an asset on the ledger
+func (s *SmartContract) ClaimAsset(ctx contractapi.TransactionContextInterface, assetAgreementBytes string, claimInfoBytes string) error {
+
+	assetAgreement := &common.AssetExchangeAgreement{}
+	err := proto.Unmarshal([]byte(assetAgreementBytes), assetAgreement)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+	// display the requested asset agreement
+	log.Info(fmt.Printf("assetExchangeAgreement: %+v\n", assetAgreement))
+
+	claimInfo := &common.AssetClaimHTLC{}
+	err = proto.Unmarshal([]byte(claimInfoBytes), claimInfo)
+	if err != nil {
+		errorMsg := fmt.Sprintf("Unmarshal error: %s", err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	// display the claim information
+	log.Info(fmt.Printf("claimInfo: %+v\n", claimInfo))
+
+	assetLockKey := assetAgreement.Type + ":" + assetAgreement.Id
+
+	assetLockValBytes, err := ctx.GetStub().GetState(assetLockKey)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+        if assetLockValBytes == nil {
+		errorMsg := fmt.Sprintf("No asset of type %s and ID %s is locked", assetAgreement.Type, assetAgreement.Id)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	assetLockVal := AssetLockValue{}
+	err = json.Unmarshal(assetLockValBytes, &assetLockVal)
+	if err != nil {
+		errorMsg := fmt.Sprintf("Unmarshal error: %s", err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	if assetLockVal.Locker != assetAgreement.Locker || assetLockVal.Recipient != assetAgreement.Recipient {
+		errorMsg := fmt.Sprintf("Can not claim asset of type %s and ID %s as it is locked by %s for %s", assetAgreement.Type, assetAgreement.Id, assetLockVal.Locker, assetLockVal.Recipient)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	// compute the hash from the preimage
+	isCorrectPreimage, err := checkIfCorrectPreimage(string(claimInfo.HashPreimage), string(assetLockVal.Hash))
+	if err != nil {
+		errorMsg := fmt.Sprintf("Claim asset of type %s and ID %s error: %v", assetAgreement.Type, assetAgreement.Id, err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	if isCorrectPreimage == false {
+		errorMsg := fmt.Sprintf("Can not claim asset of type %s and ID %s as the hash preimage is not matching", assetAgreement.Type, assetAgreement.Id)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	// Check if expiry time is elapsed
+	currentTimeSecs := time.Now().UnixNano() / int64(time.Second)
+	if uint64(currentTimeSecs) >= assetLockVal.ExpiryTimeSecs {
+		errorMsg := fmt.Sprintf("Can not claim asset of type %s and ID %s as the expiry time is already elapsed", assetAgreement.Type, assetAgreement.Id)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	err = ctx.GetStub().DelState(assetLockKey)
```

Comment:
How would a counter-party in the second network learn about the pre-image? Recall that we require the counter-party to be able to learn about the pre-image when the pre-image is first revealed on-chain in the second network so that it can then use it to claim its rightful asset on the first network (we can't assume or rely on off-chain communication here). If I am not mistaken the code above does not persist the pre-image, and in fact, removes the agreement at the end of a claim.

Reply:
Counterparty will learn about the preimage through events (not yet implemented) in the app layer. I don't see a need to persist it in the ledger. If push comes to shove, we can use a combination of history queries and block fetches to figure out what the preimage was, but it can be stored and used just in the app layer.

Reply:
Yes, the preimage is already available on the ledger of the second network.